File: CS/WindowsApplication21_Dates/WindowsApplication21/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraPivotGrid;

namespace WindowsApplication21
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
            FillDataSource(new DateTime(2010, 1, 1), new DateTime(2012, 12, 31));

        }

        private void FillDataSource(DateTime beginDate, DateTime endDate)
        {
            DataTable dt = dataSet1.Tables["Data"];
            Random r = new Random();
            int i = 0;
            for (DateTime d = beginDate; d <= endDate; d = d.AddDays(5))
            {
                dt.Rows.Add(new object[] { "Name " + i % 5, "Type " + i % 3, d, r.Next(10) });
                i++;
            }
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            pivotGridControl1.BestFit();
        }

        private void pivotGridControl1_CustomCellValue(object sender, PivotCellValueEventArgs e)
        {
            PivotGridControl pivot = sender as PivotGridControl;
            if (ReferenceEquals(e.DataField, fieldRank))
                if (e.RowField != null)
                {

                    object[] rowValues = GetRowFieldValues(e);
                    object[] columnValues = GetColumnFieldValues(e);
                    decimal currentValue = Convert.ToDecimal(e.GetCellValue(columnValues, rowValues, e.DataField));
                    List<object> visibleValues = e.RowField.GetVisibleValues();
                    int rank = 1;
                    foreach (object val in visibleValues)
                    {
                        rowValues[rowValues.Length - 1] = val;
                        object comparableValue = e.GetCellValue(columnValues, rowValues, e.DataField);
                        if (comparableValue != null && Convert.ToDecimal(comparableValue) > currentValue)
                            rank++;
                    }
                    e.Value = rank;
                }
                else
                    e.Value = null;

        }

        private object[] GetColumnFieldValues(PivotCellValueEventArgs e)
        {
            List<object> columnValues = new List<object>();
            foreach (PivotGridField field in e.GetColumnFields())
                columnValues.Add(e.GetFieldValue(field));
            return columnValues.ToArray();
        }

        private object[] GetRowFieldValues(PivotCellValueEventArgs e)
        {
            List<object> rowValues = new List<object>();
            foreach (PivotGridField field in e.GetRowFields())
                rowValues.Add(e.GetFieldValue(field));
            return rowValues.ToArray();
        }


    }
}

File: VB/WindowsApplication21/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraPivotGrid

Namespace WindowsApplication21
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
			FillDataSource(New DateTime(2010, 1, 1), New DateTime(2012, 12, 31))

		End Sub

		Private Sub FillDataSource(ByVal beginDate As DateTime, ByVal endDate As DateTime)
			Dim dt As DataTable = dataSet1.Tables("Data")
			Dim r As New Random()
			Dim i As Integer = 0
			Dim d As DateTime = beginDate
			Do While d <= endDate
				dt.Rows.Add(New Object() { "Name " & i Mod 5, "Type " & i Mod 3, d, r.Next(10) })
				i += 1
				d = d.AddDays(5)
			Loop
		End Sub

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			pivotGridControl1.BestFit()
		End Sub

		Private Sub pivotGridControl1_CustomCellValue(ByVal sender As Object, ByVal e As PivotCellValueEventArgs) Handles pivotGridControl1.CustomCellValue
			Dim pivot As PivotGridControl = TryCast(sender, PivotGridControl)
			If ReferenceEquals(e.DataField, fieldRank) Then
				If e.RowField IsNot Nothing Then

					Dim rowValues() As Object = GetRowFieldValues(e)
					Dim columnValues() As Object = GetColumnFieldValues(e)
					Dim currentValue As Decimal = Convert.ToDecimal(e.GetCellValue(columnValues, rowValues, e.DataField))
					Dim visibleValues As List(Of Object) = e.RowField.GetVisibleValues()
					Dim rank As Integer = 1
					For Each val As Object In visibleValues
						rowValues(rowValues.Length - 1) = val
						Dim comparableValue As Object = e.GetCellValue(columnValues, rowValues, e.DataField)
						If comparableValue IsNot Nothing AndAlso Convert.ToDecimal(comparableValue) > currentValue Then
							rank += 1
						End If
					Next val
					e.Value = rank
				Else
					e.Value = Nothing
				End If
			End If

		End Sub

		Private Function GetColumnFieldValues(ByVal e As PivotCellValueEventArgs) As Object()
			Dim columnValues As New List(Of Object)()
			For Each field As PivotGridField In e.GetColumnFields()
				columnValues.Add(e.GetFieldValue(field))
			Next field
			Return columnValues.ToArray()
		End Function

		Private Function GetRowFieldValues(ByVal e As PivotCellValueEventArgs) As Object()
			Dim rowValues As New List(Of Object)()
			For Each field As PivotGridField In e.GetRowFields()
				rowValues.Add(e.GetFieldValue(field))
			Next field
			Return rowValues.ToArray()
		End Function


	End Class
End Namespace